packages/integration-sdk-runtime/src/synchronization/batchBySize.ts

```diff
@@ -0,0 +1,102 @@
+import {
+  Entity,
+  IntegrationLogger,
+  IntegrationWarnEventName,
+  Relationship,
+} from '@jupiterone/integration-sdk-core';
+import { UploadDataLookup } from '.';
+import { MAX_BATCH_SIZE_IN_BYTES } from './shrinkBatchRawData';
+
+/**
+Converts an array of graphObjects into an array of groups of this objects.
+@param graphObjects  Entities or relationships to group.
+
+@param maximumBatchSizeInBytes The maximum size of the group of objects.
+
+@param logger 
+ */
+export function batchGraphObjectsBySizeInBytes<
+  T extends UploadDataLookup,
+  K extends keyof T,
+>(
+  graphObjects: T[K][],
+  maximumBatchSizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  const batches: T[K][][] = [];
+
+  let currentBatch: T[K][] = [];
+  let currentBatchSize: number = 0;
+  for (const graphObject of graphObjects) {
+    const typedGraphObject = graphObject as Entity | Relationship;
+
+    let graphObjectSizeInBytes = getSizeOfObject(graphObject);
+
+    if (graphObjectSizeInBytes > maximumBatchSizeInBytes) {
+      const rawDataSizeInBytes = typedGraphObject._rawData
+        ? getSizeOfObject(typedGraphObject._rawData)
+        : 0;
+
+      if (
+        graphObjectSizeInBytes - rawDataSizeInBytes >
+        maximumBatchSizeInBytes
+      ) {
+        logger.warn(
+          {
+            _key: typedGraphObject._key,
+            _type: typedGraphObject.type,
+            maximumBatchSizeInBytes,
+            graphObjectSizeInBytes,
+            rawDataSizeInBytes,
+          },
+          'Graph object is larger than batch size in bytes; cannot upload graph object.',
+        );
+        logger.publishWarnEvent({
+          name: IntegrationWarnEventName.IngestionLimitEncountered,
+          description:
+            'Graph object is larger than batch size in bytes; cannot upload graph object. Will skip graph object.',
+        });
```

Comment:
After this PR gets merged, let's talk with @VDubber about how best to handle these job log events and create a follow-on PR with whatever we come up with. For example:

1. What if many entities are too large? We probably want to avoid sending hundreds or thousands of these warn events to user job logs.
2. Should we include information about _what_ couldn't be uploaded? For example, the `_type`, or the `_key`? 
3. How is the user alerted to these `warn_ingestion_limit_encountered` events? Is the job marked as completed with errors, completed with some incomplete data? Is the user error messaging appropriate for these cases?